### ISSUE TITLE
null comparison fails in switch statement with `--strictNullChecks`

### ISSUE THREAD

**Issue by bmayen** [Microsoft/TypeScript#8971]
**TypeScript Version:** 

nightly (1.9.0-dev.20160604-1.0)

**Code**

```ts
let foo:number|null = 0;

switch(foo) {
  case null:
    break;
  case 0:
    console.log('bar');
}
```

**Expected behavior:**
Null check should be allowed

**Actual behavior:** 
Compiler complains "TypeScript Type 'null' is not comparable to type 'number'. (TS2678)"

**Note:**
If I add an 'if' condition before the switch, the compiler does not complain about either the 'if' or the switch. For instance:

```ts
let foo:number|null = 0;

if (foo === null) { console.log('bar'); }

switch(foo) {
  case null:
    break;
  case 0:
    console.log('bar');
}
```

**Comment by ahejlsberg** [comment]
In your first example the checker complains about the `null` case because control flow analysis has narrowed the type of `foo` to `number` and thus concluded (rightly so) that `foo` is can't be `null`. It is similar to the following:

```ts
let x: string | number;
x = "foo";  // x is known to be a string in the following
switch (x) {
    case 0:  // Error, x can't be number
    case 1:  // Error, x can't be number
    case "foo":
    case "bar":
}
```

Now, the thing that's confusing is that we allow the `foo === null` in the `if` statement in your second example. See #8452 for the reasoning on why that's permitted. We should probably extend that to switch statements as well.

The reason it works when you insert the `if` statement is that we take that statement as an *assertion* that `x` could be null at that point, even though control flow analysis has concluded it couldn't. See #8548 for the reasoning on that feature.

To sum up, the issue here is that we should always allow `null` and `undefined` cases in switch statements to be consistent with #8452.

**Comment by bmayen** [comment]
Thanks for the explanation!
